### PR TITLE
Upgrade pyperf for native stacks

### PR DESCRIPTION
**Pull request by YishaiZinkin**

### CODE REVIEW

File: scripts/pyperf_env.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+set -euo pipefail
+
+apt-get update
+DEBIAN_FRONTEND=noninteractive apt-get install -y curl git build-essential iperf llvm-9-dev libclang-9-dev \
+  cmake python3 flex bison libelf-dev libz-dev
+
+cd /tmp
+
+# Install libunwind
+curl -L http://download.savannah.nongnu.org/releases/libunwind/libunwind-1.4.0.tar.gz -o libunwind-1.4.0.tar.gz
```
Comment (Jongy) [inline scripts/pyperf_env.sh:15]: Funny, this is now the 3rd libunwind we build during gProfiler's build :)

Or actually, dumb

Comment (Jongy) [reply scripts/pyperf_env.sh:15]: Not easy to solve tho

Comment (YishaiZinkin) [reply scripts/pyperf_env.sh:15]: We're also installing the same packages over and over again 🤷🏽

File: Dockerfile [hunk 2]
```diff
@@ -28,10 +28,8 @@ RUN ./perf_build.sh
 # ubuntu 20.04
 FROM ubuntu@sha256:cf31af331f38d1d7158470e095b132acd126a7180a54f263d386da88eb681d93 AS bcc-builder
 
-RUN apt-get update
```
Comment (Jongy) [inline Dockerfile:31]: Note: also needs to be updated in `pyi.Dockerfile` which is CentOS based :/

You're now trying to share the `pyperf_env.sh` script - you can try doing so, or clone just a script `scripts/build_libunwind.sh`

Comment (YishaiZinkin) [reply Dockerfile:31]: Separating `libunwind` build to its own script makes sense

File: scripts/pyperf_env.sh (hunk 1 above)
Comment (Jongy) [reply scripts/pyperf_env.sh:15]: yup

**Comment by Jongy** [comment]
PR title can also be "upgrade pyperf for native stacks" which really is the feature here

**Comment by Jongy** [comment]
Also, nice to add some basic sanity tests. See `test_python_ebpf` which already tests for a Python function and a kernel function. Add a native function to the list

**Comment by Jongy** [comment]
Built locally:
```
[2021-07-01 09:41:06,663] DEBUG: gprofiler.python: eBPF profiler error: Command '['/app/gprofiler/resources/python/pyperf/PyPerf', '--output', '/tmp/gprofiler_tmp/tmph52m8tg9/.test', '-F', '1', '--duration', '1']' returned non-zero exit status 127. 
stdout: 
stderr: /app/gprofiler/resources/python/pyperf/PyPerf: error while loading shared libraries: libunwind-x86_64.so.8: cannot open shared object file: No such file or directory
```

**Comment by YishaiZinkin** [comment]
Whoops

**Comment by Jongy** [comment]
> Also, nice to add some basic sanity tests. See `test_python_ebpf` which already tests for a Python function and a kernel function. Add a native function to the list

Bump :)

**Review by Jongy — CHANGES_REQUESTED** [review]
Only tests, other than that - looks good

File: scripts/libunwind-container-support.patch [hunk 3]
```diff
@@ -0,0 +1,74 @@
+commit 831459ee961e7d673bbd83e40d0823227c66db33
```
Comment (Jongy) [inline scripts/libunwind-container-support.patch:1]: Did you make any changes over it? If so, which? (comment in code maybe), and if not, can replace with `curl https://github.com/../../...patch`

**Comment by Jongy** [comment]
BCC PR: https://github.com/Granulate/bcc/pull/24

File: scripts/libunwind-container-support.patch (hunk 3 above)
Comment (YishaiZinkin) [reply scripts/libunwind-container-support.patch:1]: I did

File: scripts/libunwind_build.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+set -euo pipefail
+
+curl -L http://download.savannah.nongnu.org/releases/libunwind/libunwind-1.5.0.tar.gz -o libunwind-1.5.0.tar.gz
+tar -xf libunwind-1.5.0.tar.gz
+pushd libunwind-1.5.0
+patch -p1 < ../libunwind-container-support.patch
+./configure --prefix=/usr --disable-tests --disable-documentation && make install
```
Comment (Jongy) [inline scripts/libunwind_build.sh:12]: `make -j`?

File: scripts/libunwind-container-support.patch (hunk 3 above)
Comment (YishaiZinkin) [reply scripts/libunwind-container-support.patch:1]: Actually, I can make this changes with a simple `sed`

File: scripts/libunwind_build.sh (hunk 4 above)
Comment (Jongy) [reply scripts/libunwind_build.sh:12]: I see it wasn't with `-j` before, but no reason not to make it faster :)

File: scripts/libunwind_build.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+set -euo pipefail
+
+curl -L http://download.savannah.nongnu.org/releases/libunwind/libunwind-1.5.0.tar.gz -o libunwind-1.5.0.tar.gz
```
Comment (Jongy) [inline scripts/libunwind_build.sh:8]: Can be https o.O

Comment (YishaiZinkin) [reply scripts/libunwind_build.sh:8]: Whoops. I copy-pasted it from somewhere in the code and didn't noticed

Comment (Jongy) [reply scripts/libunwind_build.sh:8]: Yeah it was this way from earlier haha. Let's improve

File: scripts/libunwind-container-support.patch (hunk 3 above)
Comment (Jongy) [reply scripts/libunwind-container-support.patch]: kk

**Comment by Jongy** [comment]
Okay, looks good. @YishaiZinkin need to merge from master & merge the BCC PR & update the tag here. Then it's good to go.

@guybortnikov please help with the phpspy issue in the tests, I won't have time till the evening.

File: Dockerfile [hunk 6]
```diff
@@ -28,10 +28,8 @@ RUN ./perf_build.sh
...
-
-RUN DEBIAN_FRONTEND=noninteractive apt-get install -y git build-essential iperf llvm-9-dev libclang-9-dev \
-  cmake python3 flex bison libelf-dev libz-dev
+COPY ./scripts/pyperf_env.sh .
```
Comment (Jongy) [inline Dockerfile]: Things started to fail from this point (first commit of the PR).

I don't get it here though, we've only changed stuff in the `bcc-builder` container, and we copy just the `PyPerf` binary....

Comment (YishaiZinkin) [reply Dockerfile]: I'll open a dummy PR to debug this

**Review by Jongy — APPROVED** [review]
Can merge once CI passes

**Comment by Jongy** [comment]
Failure is unrelated

**Pull request merged**